src/echo-common/src/cached-task.ts

```diff
@@ -3,84 +3,110 @@ import { throttleTime } from 'rxjs/operators'
 import { filterNullish } from 'ts-ratchet'
 import { EchoDirService } from './echo-dir-service'
 
-export type CachedTaskEvent<T> = {
+export type SmartCacheLoadMode = "Default" | "CacheOnly" | "NoCache"
+
+export type SmartCacheLoadConfig = {
+  key: string,
+  source: string,
+  mode?: SmartCacheLoadMode
+}
+
+export type SmartCacheResultEvent<T> = {
+  type: "result",
+  key: string,
+  result: T | null | undefined,
+  timestampMs: number
+}
+
+export type SmartCacheEvent<T> = SmartCacheResultEvent<T>
+
+export type SmartCacheLoadEvent = {
   key: string
-  state?: string
-  status?: string
-  result?: T | null | undefined
-  timestampMs?: number
 }
 
-export class CachedTask<T> {
-  private tasks$ = new Subject<CachedTaskEvent<T>>()
-  public events$ = new Subject<CachedTaskEvent<T>>()
-  public cache$ = new BehaviorSubject<{ [key: string]: CachedTaskEvent<T> }>({})
+export class SmartCache<T> {
+  private taskQueue$ = new Subject<SmartCacheLoadEvent>()
+  private events$ = new Subject<SmartCacheEvent<T>>()
+
+  //What if there was on BehaviorSubject per key that contained the state {loading, error, data}
+
+  private cache$ = new BehaviorSubject<{ [key: string]: SmartCacheEvent<T> }>({})
+  private localCacheChecked: { [key: string]: boolean } = {}
 
   constructor(
     private dir: EchoDirService,
     loadFun: (key: string) => Observable<T | null>
   ) {
-    this.tasks$
+    this.taskQueue$
       .pipe(
-        filter(({ key }) => !this.isValid(this.cache$.value[key])),
+        map((event) => this.loadFromLocalIfValid(event)),
+        filterNullish(),
         groupBy((item) => item.key),
-        mergeMap((group) => group.pipe(throttleTime(60000))),
-        map((e) => (this.loadFromLocal(e.key) ? null : e)),
+        mergeMap((group) => group.pipe(throttleTime(60000))), //can this be concat map, need some sort of switchmap interally to allow throttle bypass
```

Comment:
This makes sense, some way to bypass the throllte to trigger user-actions or internal processes, which do not rely on the internal throttle/intervals